scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,1009 @@
+        // timeout is ignored inside the method implementation, it is handled by the runActivity method
+        return waitUntil(originator, taskid, channel, key, predicate);
+    }
+
+    @Override
+    public PredicateActionResult waitUntilThen(String originator, TaskId taskid, String channel, String key,
+            String predicate, String thenRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException {
+        PredicateActionResult answer;
+        try {
+            Channel chosenChannel = getChannel(channel);
+            Serializable value = chosenChannel.get(key);
+            if (evaluateClosure(predicate, BiPredicate.class).test(key, value)) {
+                answer = new PredicateActionResult(true,
+                                                   chosenChannel.compute(key,
+                                                                         (BiFunction<? super String, ? super Serializable, ? extends Serializable>) evaluateClosure(thenRemappingFunction,
+                                                                                                                                                                    BiFunction.class)));
+            } else {
+                throw new IllegalStateException("Predicate " + predicate + " is not met");
+            }
+
+            logWithContextAndPersist(taskid,
+                                     channel,
+                                     "Run waitUntil" + ON_KEY + QUOTE + key + QUOTE + RESULT_IS + answer,
+                                     Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+        return answer;
+    }
+
+    @Override
+    public PredicateActionResult waitUntilThen(String originator, TaskId taskid, String channel, String key,
+            String predicate, long timeout, String thenRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException, TimeoutException {
+        // timeout is ignored inside the method implementation, it is handled by the runActivity method
+        return waitUntilThen(originator, taskid, channel, key, predicate, thenRemappingFunction);
+    }
+
+    private void logWithContextAndPersist(TaskId taskId, String channel, String message, Level level)
+            throws IOException, InvalidChannelException {
+        logWithContext(taskId, channel, message, null, level);
+        try {
+            commitIfNeeded(channel);
+        } catch (IOException e) {
+            logWithContext(taskId, channel, "Error when persisting channel content", e, Level.ERROR);
+            throw e;
+        }
+    }
+
+    private void logWithContext(TaskId taskId, String channel, String message, Throwable exception, Level level)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)